openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -197,38 +443,115 @@ class ParameterHandler(object):
 
     """
 
-    # TODO: Keep track of preferred units for parameter handlers.
-
-    # TODO: Remove these?
     _TAGNAME = None  # str of section type handled by this ParameterHandler (XML element name for SMIRNOFF XML representation)
-    _VALENCE_TYPE = None  # ChemicalEnvironment valence type string expected by SMARTS string for this Handler
-    _INFOTYPE = None  # container class with type information that will be stored in self._types
+    _INFOTYPE = None  # container class with type information that will be stored in self._parameters
     _OPENMMTYPE = None  # OpenMM Force class (or None if no equivalent)
     _DEPENDENCIES = None  # list of ParameterHandler classes that must precede this, or None
-    _DEFAULTS = {}  # dict of attributes and their default values at tag-level
-    _KWARGS = []  # list of keyword arguments accepted by the force Handler on initialization
+    _REQUIRED_SPEC_ATTRIBS = [] # list of kwargs that must be present during handler initialization
+    _DEFAULT_SPEC_ATTRIBS = {}  # dict of tag-level attributes and their default values
+    _OPTIONAL_SPEC_ATTRIBS = []  # list of non-required attributes that can be defined on initialization
+    _INDEXED_ATTRIBS = []  # list of parameter attribs that will have consecutive numerical suffixes starting at 1
+    _REQUIRE_UNITS = {}  # dict of {header attrib : unit } for input checking
+    _ATTRIBS_TO_TYPE = {} # dict of attribs that must be cast to a specific type to be interpreted correctly
+    _KWARGS = [] # Kwargs to catch when create_force is called
     _SMIRNOFF_VERSION_INTRODUCED = 0.0  # the earliest version of SMIRNOFF spec that supports this ParameterHandler
     _SMIRNOFF_VERSION_DEPRECATED = None  # if deprecated, the first SMIRNOFF version number it is no longer used
-    _REQUIRE_UNITS = dict(
-    )  # dict of parameters that require units to be defined
 
-    # TODO: Do we need to store the parent forcefield object?
-    def __init__(self, forcefield, **kwargs):
+
+    def __init__(self, permit_cosmetic_attributes=False, **kwargs):
         """
+        Initialize a ParameterHandler, optionally with a list of parameters and other kwargs.
 
         Parameters
         ----------
-        forcefield : openforcefield.typing.engines.smirnoff.ForceField
+        permit_cosmetic_attributes : bool
+            Whether to accept non-spec kwargs
+        **kwargs : dict
+            The dict representation of the SMIRNOFF data source
```

Comment:
So I've been having second thoughts about this (apologies for my original suggestion!), and I'm now wondering if we should have this signature in a `from_dict(permit_cosmetic_attributes=False. **kwargs)` static constructor, and keep a classic separate construct with a signature like `__init__(parameter_types=None, permit_cosmetic_attributes=False, ..., **cosmetic_attributes)` for consistency with the rest of the API.

Reply:
I like that. So, the `from_dict` function would attach units and perform other default behaviors for preparing OFFXML data, and then pass the formatted results to `__init__`

Reply:
In the interest of conciseness, I'm going to treat this as an issue separate from this PR.